File: acra-core/src/main/java/org/acra/util/ProcessFinisher.java
/*
 * Copyright (c) 2016
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.acra.util;

import android.app.Activity;
import android.app.ActivityManager;
import android.content.Context;
import android.content.Intent;
import android.os.Process;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.acra.ACRA;
import org.acra.builder.LastActivityManager;
import org.acra.config.CoreConfiguration;
import org.acra.sender.JobSenderService;
import org.acra.sender.LegacySenderService;

import java.util.List;

import static org.acra.ACRA.LOG_TAG;

/**
 * Takes care of cleaning up a process and killing it.
 *
 * @author F43nd1r
 * @since 4.9.2
 */

public final class ProcessFinisher {
    private final Context context;
    private final CoreConfiguration config;
    private final LastActivityManager lastActivityManager;

    public ProcessFinisher(@NonNull Context context, @NonNull CoreConfiguration config, @NonNull LastActivityManager lastActivityManager) {
        this.context = context;
        this.config = config;
        this.lastActivityManager = lastActivityManager;
    }

    public void endApplication() {
        stopServices();
        killProcessAndExit();
    }

    public void finishLastActivity(@Nullable Thread uncaughtExceptionThread) {
        if (ACRA.DEV_LOGGING) ACRA.log.d(LOG_TAG, "Finishing activities prior to killing the Process");
        boolean wait = false;
        for(Activity activity : lastActivityManager.getLastActivities()) {
            final boolean isMainThread = uncaughtExceptionThread == activity.getMainLooper().getThread();
            final Runnable finisher = () -> {
                activity.finish();
                if (ACRA.DEV_LOGGING) ACRA.log.d(LOG_TAG, "Finished " + activity.getClass());
            };
            if (isMainThread) {
                finisher.run();
            } else {
                // A crashed activity won't continue its lifecycle. So we only wait if something else crashed
                wait = true;
                activity.runOnUiThread(finisher);
            }
        }
        if (wait) {
            lastActivityManager.waitForAllActivitiesDestroy(100);
        }
        lastActivityManager.clearLastActivities();
    }

    private void stopServices() {
        if (config.stopServicesOnCrash()) {
            try {
                final ActivityManager activityManager = SystemServices.getActivityManager(context);
                final List<ActivityManager.RunningServiceInfo> runningServices = activityManager.getRunningServices(Integer.MAX_VALUE);
                final int pid = Process.myPid();
                for (ActivityManager.RunningServiceInfo serviceInfo : runningServices) {
                    if (serviceInfo.pid == pid && !LegacySenderService.class.getName().equals(serviceInfo.service.getClassName()) && !JobSenderService.class.getName().equals(serviceInfo.service.getClassName())) {
                        try {
                            final Intent intent = new Intent();
                            intent.setComponent(serviceInfo.service);
                            context.stopService(intent);
                        } catch (SecurityException e) {
                            if (ACRA.DEV_LOGGING) ACRA.log.d(LOG_TAG, "Unable to stop Service " + serviceInfo.service.getClassName() + ". Permission denied");
                        }
                    }
                }
            } catch (SystemServices.ServiceNotReachedException e) {
                ACRA.log.e(LOG_TAG, "Unable to stop services", e);
            }
        }
    }

    private void killProcessAndExit() {
        Process.killProcess(Process.myPid());
        System.exit(10);
    }
}

File: acra-core/src/main/java/org/acra/util/ProcessFinisher.kt
/*
 * Copyright (c) 2016
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.acra.util

import android.content.Context
import android.content.Intent
import android.os.Process
import org.acra.builder.LastActivityManager
import org.acra.config.CoreConfiguration
import org.acra.log.debug
import org.acra.log.error
import org.acra.sender.JobSenderService
import org.acra.sender.LegacySenderService
import org.acra.util.SystemServices.ServiceNotReachedException
import org.acra.util.SystemServices.getActivityManager
import kotlin.system.exitProcess

/**
 * Takes care of cleaning up a process and killing it.
 *
 * @author F43nd1r
 * @since 4.9.2
 */
class ProcessFinisher(private val context: Context, private val config: CoreConfiguration, private val lastActivityManager: LastActivityManager) {
    fun endApplication() {
        stopServices()
        killProcessAndExit()
    }

    fun finishLastActivity(uncaughtExceptionThread: Thread?) {
        debug { "Finishing activities prior to killing the Process" }
        var wait = false
        for (activity in lastActivityManager.lastActivities) {
            val finisher = Runnable {
                activity.finish()
                debug { "Finished ${activity.javaClass}" }
            }
            if (uncaughtExceptionThread === activity.mainLooper.thread) {
                finisher.run()
            } else {
                // A crashed activity won't continue its lifecycle. So we only wait if something else crashed
                wait = true
                activity.runOnUiThread(finisher)
            }
        }
        if (wait) {
            lastActivityManager.waitForAllActivitiesDestroy(100)
        }
        lastActivityManager.clearLastActivities()
    }

    @Suppress("DEPRECATION")
    private fun stopServices() {
        if (config.stopServicesOnCrash) {
            try {
                val activityManager = getActivityManager(context)
                val runningServices = activityManager.getRunningServices(Int.MAX_VALUE)
                val pid = Process.myPid()
                for (serviceInfo in runningServices) {
                    if (serviceInfo.pid == pid && LegacySenderService::class.java.name != serviceInfo.service.className
                            && JobSenderService::class.java.name != serviceInfo.service.className) {
                        try {
                            val intent = Intent()
                            intent.component = serviceInfo.service
                            context.stopService(intent)
                        } catch (e: SecurityException) {
                            debug { "Unable to stop Service ${serviceInfo.service.className}. Permission denied" }
                        }
                    }
                }
            } catch (e: ServiceNotReachedException) {
                error(e) { "Unable to stop services" }
            }
        }
    }

    private fun killProcessAndExit() {
        Process.killProcess(Process.myPid())
        exitProcess(10)
    }
}